File: Longest Balanced Substring I/Longest Balanced Substring I.py
# Solution for Longest Balanced Substring I in PY

class Solution:
    def longestBalanced(self, s: str) -> int:

        n = len(s)
        max_length = 0

        for start in range(n):

            frequency = [0] * 26

            for end in range(start, n):

                frequency[ord(s[end]) - ord('a')] += 1

                if self.is_balanced(frequency):
                    max_length = max(max_length, end - start + 1)

        return max_length

    def is_balanced(self, freq):

        required_frequency = 0

        for count in freq:

            if count == 0:
                continue

            if required_frequency == 0:
                required_frequency = count
            elif count != required_frequency:
                return False

        return True

File: Longest Balanced Substring I/Longest Balanced Substring I.java
// Solution for Longest Balanced Substring I in JAVA

class Solution {

    public int longestBalanced(String s) {

        int n = s.length();
        int maxLength = 0;

        // Fix starting index
        for (int start = 0; start < n; start++) {

            int[] frequency = new int[26];

            for (int end = start; end < n; end++) {

                // Update frequency
                frequency[s.charAt(end) - 'a']++;

                if (isBalanced(frequency)) {
                    maxLength = Math.max(maxLength, end - start + 1);
                }
            }
        }

        return maxLength;
    }

    // Helper method
    private boolean isBalanced(int[] freq) {

        int requiredFrequency = 0;

        for (int i = 0; i < 26; i++) {

            if (freq[i] == 0)
                continue;

            if (requiredFrequency == 0) {
                requiredFrequency = freq[i];
            } else if (freq[i] != requiredFrequency) {
                return false;
            }
        }

        return true;
    }
}
